### PR TITLE
flow: implement discovery.dns

### LINKED ISSUE #2265: Flow: discovery.dns component
For feature parity, we should bring over the upstream Prometheus SD mechanism and introduce it as a Flow component.

### PR DESCRIPTION
**Pull request by captncraig**
#### Which issue(s) this PR fixes

Fixes #2265

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [ ] Tests updated

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -46,6 +46,7 @@ Main (unreleased)
   - `otelcol.exporter.loki` forwards OTLP-formatted data to compatible `loki`
     receivers. (@tpaschalis)
   - `loki.source.gelf` listens for Graylog logs. (@mattdurham)
+  - `discovery.dns` discivers targets from dns records (@captncraig)
```
Comment (jcreixell) [inline CHANGELOG.md:49]:
```suggestion
  - `discovery.dns` discovers targets from dns records (@captncraig)
```

**Review by jcreixell — COMMENTED** [review]
a few unit tests for `UnmarshalRiver` and `Convert` functions would be nice, apart from that LGTM

File: docs/sources/flow/reference/components/discovery.dns.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
+---
+aliases:
+- /docs/agent/latest/flow/reference/components/discovery.dns
+title: discovery.dns
+---
+
+# discovery.dns
+
+`discovery.dns` discovers scrape targets from dns records.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/discovery.dns.md:9]:
```suggestion
`discovery.dns` discovers scrape targets from DNS records.
```

File: docs/sources/flow/reference/components/discovery.dns.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+```
+
+## Arguments
+
+The following arguments are supported:
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`names` | `list(string)` | DNS names to look up. | | yes
+`port` | `number` | Port to use for collecting metrics. Not used for SRV records | `0` | no
```
Comment (karengermond) [inline docs/sources/flow/reference/components/discovery.dns.md:26]:
```suggestion
`port` | `number` | Port to use for collecting metrics. Not used for SRV records. | `0` | no
```

File: docs/sources/flow/reference/components/discovery.dns.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+`refresh_interval` | `duration` | How often to query dns for updates. | `"30s"` | no
```
Comment (karengermond) [inline docs/sources/flow/reference/components/discovery.dns.md:27]:
```suggestion
`refresh_interval` | `duration` | How often to query DNS for updates. | `"30s"` | no
```

File: docs/sources/flow/reference/components/discovery.dns.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+`type` | `string` | Type of dns record to query. Must be one of SRV, A, AAAA, or MX | `"SRV"` | no
```
Comment (karengermond) [inline docs/sources/flow/reference/components/discovery.dns.md:28]:
```suggestion
`type` | `string` | Type of DNS record to query. Must be one of SRV, A, AAAA, or MX. | `"SRV"` | no
```

File: docs/sources/flow/reference/components/discovery.dns.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+## Exported fields
+
+The following fields are exported and can be referenced by other components:
```
Comment (karengermond) [inline docs/sources/flow/reference/components/discovery.dns.md:32]:
```suggestion
The following field is exported and can be referenced by other components:
```

File: docs/sources/flow/reference/components/discovery.dns.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+Name | Type | Description
+---- | ---- | -----------
+`targets` | `list(map(string))` | The set of targets discovered from the docker API.
+
+Each target includes the following labels:
+
+* `__meta_dns_name`: the record name that produced the discovered target.
+* `__meta_dns_srv_record_target`: the target field of the SRV record
+* `__meta_dns_srv_record_port`: the port field of the SRV record
+* `__meta_dns_mx_record_target`: the target field of the MX record
```
Comment (karengermond) [inline docs/sources/flow/reference/components/discovery.dns.md:43]:
```suggestion
* `__meta_dns_name`: Name of the record that produced the discovered target.
* `__meta_dns_srv_record_target`: Target field of the SRV record.
* `__meta_dns_srv_record_port`: Port field of the SRV record.
* `__meta_dns_mx_record_target`: Target field of the MX record.
```

File: docs/sources/flow/reference/components/discovery.dns.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+`discovery.dns` does not expose any component-specific debug information.
+
+### Debug metrics
+
+`discovery.dns` does not expose any component-specific debug metrics.
+
+## Examples
+
+This simple example discovers targets from an A record
```
Comment (karengermond) [inline docs/sources/flow/reference/components/discovery.dns.md:62]:
```suggestion
This example discovers targets from an A record.
```

**Review by karengermond — APPROVED** [review]
LGTM!

File: component/discovery/dns/dns.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+func (args *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*args = DefaultArguments
+
+	type arguments Arguments
+	if err := f((*arguments)(args)); err != nil {
+		return err
+	}
+	switch strings.ToUpper(args.Type) {
+	case "SRV":
+	case "A", "AAAA", "MX":
```
Comment (jcreixell) [inline component/discovery/dns/dns.go:51]: let's add a test for these cases

Comment (jcreixell) [reply component/discovery/dns/dns.go:51]: and another one for a valid config to make sure the parsing works fine

**Pull request merged**